dev/server.ts

```diff
@@ -0,0 +1,149 @@
+import * as fs from 'fs';
+import * as os from 'os';
+import * as path from 'path';
+
+import * as babel from '@babel/core';
+import * as chokidar from 'chokidar';
+import * as express from 'express';
+import * as morgan from 'morgan';
+import * as portfinder from 'portfinder';
+
+import { makeListing } from '../src/common/tools/crawl.js';
+
+// Import the project's babel.config.js. We'll use the same config for the runtime compiler.
+const babelConfig = {
+  ...require(path.resolve(__dirname, '../babel.config.js'))({
+    cache: () => {
+      /* not used */
+    },
+  }),
+  sourceMaps: 'inline',
+};
+
+// Caches for the generated listing file and compiled TS sources to speed up reloads.
+const listingCache = new Map<string, string>();
+const compileCache = new Map<string, string>();
+
+const srcDir = path.resolve(__dirname, '../src');
+
+console.log('Watching changes in', srcDir);
+const watcher = chokidar.watch(srcDir, {
+  persistent: true,
+});
+
+// Handler to dirty the compile cache for changed .ts files.
+function dirtyCompileCache(absPath: string, stats?: fs.Stats) {
+  const relPath = path.relative(srcDir, absPath);
+  if ((stats === undefined || stats.isFile()) && relPath.endsWith('.ts')) {
+    const tsUrl = path.sep + relPath;
+    if (compileCache.has(tsUrl)) {
+      console.debug('Dirtying compile cache', tsUrl);
+    }
+    compileCache.delete(tsUrl);
+  }
+}
+
+// Handler to dirty the listing cache for directory changes. Also dirties
+// the compile cache for changed files.
```

Comment:
nit: I think the listing only cares about the `.spec.ts` files so we can dirty the listing only in changes of such files. Also does the listing list the tests inside each `.spec.ts` or just the list of all the `.spec.ts`? If it's the former I think we also need to dirty the listing cache on `change`.

Reply:
Indeed it only cares about dirs and .spec.ts. I added a condition for that.

It doesn't care about the contents of the `.spec.ts` except for the `description`. I think we're probably okay leaving that outdated..? Updating efficiently would be more involved since compiling the listing takes seconds.

Reply:
Ah yeah if the listing only cares about the description I agree that it's not worth forcing a listing recompile that takes seconds, and the description being outdated will be a small gotcha.

Reply:
I was thinking about taking the description out of the listing anyway. I am pretty sure it's not needed.

Reply:
Any more comments Kai, or shall we land this? :)